### ISSUE TITLE
readNetFromTensorflow: Unknown layer type Shape in op dropout/Shape

### ISSUE THREAD

**Issue by gzzhang** [opencv/opencv#9536]
##### System information (version)

- OpenCV =>  3.3.0
- Operating System / Platform => Ubuntu 16
- Compiler => :grey_question:

##### Detailed description

I had trained a model for MNIST with tensorflow1.3.0, the model contains a dropout layout.

`keep_prob = tf.placeholder(tf.float32, name='keep_prob');`
`h_fc1 = tf.nn.dropout(h_fc1, keep_prob);`

Then I use the model like below

`mNet = cv2.dnn.readNetFromTensorflow(model)`

The error is: 

OpenCV Error: Unspecified error (Unknown layer type Shape in op dropout/Shape) in populateNet

=========

If the model not contain dropout layer, everything is ok.

**Comment by dkurt** [comment]
@gzzhang, thanks! That looks really challenging. Even with `isTraining` flag of `tf.layers.dropout` or `tf.contrib.layers.dropout`, resulting graph has ops like `Switch` depends on the flag value.
I've tried to remove dropout nodes manually or using Tensorboard GUI but nothing.
The easiest way - save chackpoint for a whole graph:

```python
import tensorflow as tf
import numpy as np

sess = tf.Session()

# Build
inp = tf.placeholder(tf.float32, [1, 10, 10, 3], 'input')
conv = tf.layers.conv2d(inputs=inp, filters=3, kernel_size=[1, 1])
dropout = tf.layers.dropout(conv, rate=0.1, training=True)
relu = tf.nn.relu(dropout)

# Run
sess.run(tf.global_variables_initializer())

inputData = np.random.standard_normal(inp.shape).astype(np.float32)
probData = np.random.uniform(0.0, 1.0, inp.shape).astype(np.float32)
outputData = sess.run(relu, feed_dict={inp: inputData})

# Save
saver = tf.train.Saver()
saver.save(sess, 'dropout.ckpt')
```

But graph definition for graph without dropout:
```python
import tensorflow as tf
import numpy as np

sess = tf.Session()

# Build
inp = tf.placeholder(tf.float32, [1, 10, 10, 3], 'input')
conv = tf.layers.conv2d(inputs=inp, filters=3, kernel_size=[1, 1])
# dropout = tf.layers.dropout(conv, rate=0.1, training=True)
relu = tf.nn.relu(conv)
tf.train.write_graph(sess.graph.as_graph_def(), "", "dropout.pb")
```

Then, as usual, call `freeze_graph.py` tool:
```bash
python ~/tensorflow/tensorflow/python/tools/freeze_graph.py \
  --input_graph=dropout.pb \
  --input_checkpoint=dropout.ckpt \
  --output_graph=dropout_frozen.pb \
  --output_node_names=Relu
```
And `optimize_for_inference.py` tool:
```bash
python ~/tensorflow/tensorflow/python/tools/optimize_for_inference.py \
  --input dropout_frozen.pb \
  --output dropout_opt.pb \
  --frozen_graph True \
  --input_names input \
  --output_names Relu
```

**Comment by stereomatchingkiss** [comment]
@dkurt From this post, do you mean opencv3.3 do not support dropout layer of tensorflow yet?

**Comment by dkurt** [comment]
@stereomatchingkiss, DNN's TensorFlow importer designed to process models where an every layer represented as a single op with a some number of `Const` ops passed as inputs (weights, hyperparameters). In example, 
```python
inp = tf.placeholder(tf.float32, [1, 6, 5, 3], 'input')
conv = tf.layers.conv2d(inputs=inp, filters=3, kernel_size=[1, 1],
                        activation=tf.nn.relu,
                        bias_initializer=tf.random_normal_initializer())
```
<details>
 <summary>Corresponding graph definition</summary>

```
node {
  name: "input"
  op: "Placeholder"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
}
node {
  name: "conv2d/kernel"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_FLOAT
        tensor_shape {
          dim {
            size: 1
          }
          dim {
            size: 1
          }
          dim {
            size: 3
          }
          dim {
            size: 3
          }
        }
        tensor_content: "@\263\014\276\324@Y\277\\\006X?\334\253y?\214\345=?\300\353<\276H\251T\277\260G\246\276\334\375\007\277"
      }
    }
  }
}
node {
  name: "conv2d/bias"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_FLOAT
        tensor_shape {
          dim {
            size: 3
          }
        }
        tensor_content: "]e\306?\375\231\004\300\346\037\350>"
      }
    }
  }
}
node {
  name: "conv2d/convolution"
  op: "Conv2D"
  input: "input"
  input: "conv2d/kernel"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "data_format"
    value {
      s: "NHWC"
    }
  }
  attr {
    key: "padding"
    value {
      s: "VALID"
    }
  }
  attr {
    key: "strides"
    value {
      list {
        i: 1
        i: 1
        i: 1
        i: 1
      }
    }
  }
  attr {
    key: "use_cudnn_on_gpu"
    value {
      b: true
    }
  }
}
node {
  name: "conv2d/BiasAdd"
  op: "BiasAdd"
  input: "conv2d/convolution"
  input: "conv2d/bias"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "data_format"
    value {
      s: "NHWC"
    }
  }
}
node {
  name: "conv2d/Relu"
  op: "Relu"
  input: "conv2d/BiasAdd"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
}

```
</details>

The thing is that some of layers require to do more effort to fuse it in a single node. In example, Batch normalization has a set of arithmetic operations over constants by default. So even in TensorFlow runtime it computes the same values again and again just for multiply input and shift it. Different ops require different simplification approaches. Sometimes `optimize_for_inference.py` tool may help. Sometimes we should use a one and only layer implementation across different packages (`tf.nn`, `tf.layers`, `tf.contrib.layers` etc.). Sometimes we need to do manual graph transformations.

Returning back to dropout, we need to turn it off somehow in inference time. But even using the switch flag, the graph will has a set of ops to maintain it. The problem is that it's not a single op like "Dropout" with constant boolean input. I tried to fuse them or remove them but nothing. You may find similar issues like `How to drop dropout in tensorflow`, `remove dropout tensorflow`.

My answer is DNN can't support TensorFlow dropout procedure (I mean a procedure but not a layer because it's a huge subgraph) until we find a way to represent it as a single graph node.

**Comment by stereomatchingkiss** [comment]
@dkurt I do not know how to solve it, but I know this is a very difficult problem to solve, thanks for the efforts you put into dnn module.

**Comment by dkurt** [comment]
@gzzhang, @stereomatchingkiss. We've found a way to import graphs that contain dropout layers. Please, see PR https://github.com/opencv/opencv/pull/9673.
